### DISCUSSION TITLE
SQL NOT operator

### DISCUSSION THREAD

**Discussion by Maverick-501** [Blazebit/blaze-persistence#1778] [Q&A] [answered]
Hello,
I am looking for a way to express a WHERE clause like this with blaze-persistence.

`WHERE NOT compositePredicate`

and the compositePredicate being something like this which is build using the .whereOr() and .whereAnd() builder.

`compositePredicate = ((predicate1 AND predicate2) OR (predicate1 AND predicate3)) `

I know that I can negate the predicates itself e.g. using a .ne() instead of .eq() but then I would need to swap the AND/OR operators inside the compositePredicate. Since I need the compositePredicate in another query too I would be nice to be able to simply negate the whole compositePredicate. With jakarta.persistence.criteria.CriteriaBuilder I can implement this with
 
`cb.not(compositePredicate)`

I checked the documentation and disscussions here on github but couldnt find a way to negate a predicate which is build using the .whereOr() builder. 

Am I missing something or is this not possible with blaze-persistence?
Thanks!

**Comment by beikov** [comment, marked as answer]
You are right that this is not possible yet. This is a nice new feature for version 2, which could be easily implemented with the generic predicate builder: https://github.com/Blazebit/blaze-persistence/issues/1596

Possible syntax:

```
cb.where()
  .not()
    ....
  .endNot()
```

**Reply by beikov** [reply]
In fact, we could already introduce this `NegatedBuilder not()` on our `PredicateBuilder`, but we'd really only want to add it there.

**Reply by beikov** [reply]
Ok, I created https://github.com/Blazebit/blaze-persistence/issues/1779 to track this.
